docs/cookbook/publishing_with_workflow.md

```diff
@@ -0,0 +1,411 @@
+        });
+
+        $tos = [];
+
+        return array_reduce($workflowTransitions, function ($tos, Transition $item) {
+            return array_merge($tos, $item->getTos());
+        }, $tos);
+    }
+}
+```
+
+The subscriber must be registered as a service:
+
+``` yaml hl_lines="5"
+services:
+    # ...
+    AppBundle\Event\Workflow\PublishOnLastStageSubscriber:
+        arguments:
+            $publishOnLastStageWorkflows: '%app.workflow.publish_on_last_stage%'
+```
+
+You must also provide the identifier of the workflow you want the subscriber to apply to.
+Do it in the `app.workflow.publish_on_last_stage` parameter:
+
+``` yaml
+parameters:
+    app.workflow.publish_on_last_stage: [editorial_workflow, legal_workflow]
+```
+
+## Finish workflow for published content
+
+With proper permissions, you can publish content even before it has gone through a whole workflow.
+Afterwards it will still be visible in the review queue
+and in the relevant stage of **Content item(s) under review** tab under **Workflow** in the Admin panel.
+
+To avoid cluttering the tables with published content, you can use an event subscriber
+which will automatically move content to the last stage of the workflow after it has been published:
+
+``` php hl_lines="1 12"
+public static function getSubscribedEvents(): array
+{
+    return [
+        MVCEvents::API_SIGNAL => ['onPublishVersionSignal', -50],
+    ];
+}
+
+public function onPublishVersionSignal(SignalEvent $event): void
+{
+    $signal = $event->getSignal();
+
+    if (!$signal instanceof PublishVersionSignal) {
+        return;
+    }
+
+    $this->doEndWorkflows((int)$signal->contentId, (int)$signal->versionNo);
+}
+```
+
+The `doEndWorkflows()` function applies all transitions that are needed to bring the Content item
+to the final workflow stage.
+
+``` php  hl_lines="21 22"
+private function doEndWorkflows(int $contentId, int $versionNo): void
+{
+    $content = $this->contentService->loadContent($contentId, [], $versionNo);
+    $supportedWorkflows = $this->workflowRegistry->getSupportedWorkflows($content);
+
+    foreach ($supportedWorkflows as $workflow) {
+        try {
+            $workflowMetadata = $this->workflowService->loadWorkflowMetadataForContent($content, $workflow->getName());
+        } catch (NotFoundException $e) {
+            continue;
+        }
+
+        $workflowCurrentState = !empty($workflowMetadata->markings) ? end($workflowMetadata->markings)->name : '';
+        $transitions = $workflowMetadata->workflow->getDefinition()->getTransitions();
+        $lastStage = $this->getLastStage($transitions, $workflow->getName());
+
+        if (!$lastStage) {
+            continue;
+        }
+
+        $transitionsToMake = $this->findPathToLastStage($transitions, $workflowCurrentState, $lastStage);
+        $this->applyTransitions($transitionsToMake, $workflowMetadata);
+    }
+}
+```
+
+??? note "Full `EndWorkflowSubscriber` code"
```

Comment:
same as above, maybe drop-down, as the last part about registering subscriber as a service does not connect well with beginning paragraphs